### DISCUSSION TITLE
Pyarmor CI/CD License Portability and Build Process

### DISCUSSION THREAD

**Discussion by Cosmicoppai** [dashingsoft/pyarmor#1972] [Q&A] [answered]
I'm considering purchasing the Pyarmor CI/CD license for our build pipeline. Currently thinking about building in GitHub Actions but may need flexibility for future infrastructure changes. Have few questions:

1. Does the CI/CD license ($90/year) support moving between different CI/CD environments? For example:
   - From GitHub Actions to self-hosted runners
   - From one cloud provider (e.g., GitHub) to another (e.g., AWS)

2. Are there any limitations or considerations regarding:
   - Number of separate builds we can perform
   - Need for re-registration when changing CI/CD providers
   - Any specific configuration needed when switching between environments

3. Customer-Specific Build Requirements:
   - Is it possible to generate independent builds with customer-specific parameters?
   - Can we tie specific builds to individual customers using unique metadata or parameters?
   - How can we ensure builds for one customer remain separate and independent from others?
   - What are the recommended practices for managing multiple customer-specific builds in a CI/CD pipeline?

**Comment by jondy** [comment, marked as answer]
CI License doesn't care different CI/CD environments

But it will not work in these cases:
1. after expired date
2. if install new pyarmor version in the CI/CD pipeline, not same as pyarmor verion which used to request CI license
   for example, use Pyarmor 9.0.1 to request CI License, in CI/CD pipeline,   just install 9.0.x, if install 9.1.0+, it may not work
3. build rate limitation 
4. do not request new CI license if it still works

About different customer different configuration, may be `pyarmor cfg outerkey_name` works

Please refer to Pyarmor doc to understand Pyarmor more features
http://pyarmor.readthedocs.io/en/latest/
